File: harness/oracle/cases/core-whisper.ts
/**
 * Oracle case (chat-orchestration wave 1.4): the Aurora Core whisper cadence
 * trigger.
 *
 * Drives the REAL pure function from v4's
 * lib/chat/context/core-whisper-trigger.ts: shouldFireCoreWhisper (which
 * internally exercises the private isVisibleConversationalTurn and
 * findLatestContextTransitionIndex helpers — not separately exported, so
 * covered only through the public function).
 *
 * Run from inside the server checkout:
 *   cd ~/source/quilltap-server
 *   npx tsx ~/source/quilltap-v5/harness/oracle/cases/core-whisper.ts \
 *     > /tmp/oracle-core-whisper.ndjson
 */

import {
  shouldFireCoreWhisper,
  type ShouldFireCoreWhisperOptions,
} from '@/lib/chat/context/core-whisper-trigger';
import type { ChatEvent, MessageEvent } from '@/lib/schemas/chat.types';

interface Row {
  id: string;
  options: {
    events: Array<Record<string, unknown>>;
    respondingParticipantId: string;
    isContinue: boolean;
    isNudge: boolean;
    interval: number;
    silenceThreshold: number;
    fireOnContextTransition?: boolean;
  };
  out: ReturnType<typeof shouldFireCoreWhisper>;
}

const rows: Row[] = [];

const P1 = 'p-1111-participant-one';
const P2 = 'p-2222-participant-two';
const P3 = 'p-3333-participant-three';

let seq = 0;
function msg(overrides: Partial<MessageEvent> = {}): MessageEvent {
  seq += 1;
  return {
    type: 'message',
    id: `m-${seq}`,
    role: 'ASSISTANT',
    content: `content-${seq}`,
    attachments: [],
    createdAt: new Date(0).toISOString(),
    ...overrides,
  } as MessageEvent;
}

function contextSummary(): ChatEvent {
  seq += 1;
  return {
    type: 'context-summary',
    id: `cs-${seq}`,
    content: 'summary text',
    createdAt: new Date(0).toISOString(),
  } as unknown as ChatEvent;
}

function systemEvent(): ChatEvent {
  seq += 1;
  return {
    type: 'system',
    id: `sys-${seq}`,
    kind: 'note',
    createdAt: new Date(0).toISOString(),
  } as unknown as ChatEvent;
}

// Aurora's own core-whisper message targeting a given participant.
function coreWhisperMsg(targetId: string): MessageEvent {
  return msg({
    role: 'ASSISTANT',
    systemSender: 'aurora',
    systemKind: 'core-whisper',
    targetParticipantIds: [targetId],
    content: 'A quiet packet arrives.',
  });
}

function myTurn(participantId: string, content = 'my turn content'): MessageEvent {
  return msg({ role: 'ASSISTANT', participantId, content });
}

function theirTurn(participantId: string, content = 'their turn content'): MessageEvent {
  return msg({ role: 'ASSISTANT', participantId, content });
}

function userTurn(content = 'user says hi'): MessageEvent {
  return msg({ role: 'USER', participantId: null, content });
}

function staffWhisper(systemSender: MessageEvent['systemSender'] = 'lantern'): MessageEvent {
  return msg({ role: 'ASSISTANT', systemSender, content: 'staff announcement' });
}

function librarianFold(
  systemKind: string,
  targetParticipantIds?: string[] | null,
): MessageEvent {
  return msg({
    role: 'ASSISTANT',
    systemSender: 'librarian',
    systemKind,
    content: 'summary fold whisper',
    targetParticipantIds: targetParticipantIds ?? null,
  });
}

function silentMsg(participantId: string): MessageEvent {
  return msg({ role: 'ASSISTANT', participantId, isSilentMessage: true, content: 'silent' });
}

function emptyContentMsg(participantId: string): MessageEvent {
  return msg({ role: 'ASSISTANT', participantId, content: '   ' });
}

function privateWhisper(targetIds: string[], content = 'private aside'): MessageEvent {
  return msg({ role: 'ASSISTANT', targetParticipantIds: targetIds, content });
}

type Case = [string, Omit<ShouldFireCoreWhisperOptions, 'events'> & { events: ChatEvent[] }];

const cases: Case[] = [
  // ---- isContinue / isNudge short-circuit ---------------------------------
  [
    'continue-skips-entirely',
    {
      events: [],
      respondingParticipantId: P1,
      isContinue: true,
      isNudge: false,
      interval: 5,
      silenceThreshold: 3,
    },
  ],
  [
    'nudge-skips-entirely',
    {
      events: [myTurn(P1), theirTurn(P2)],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: true,
      interval: 5,
      silenceThreshold: 3,
    },
  ],
  [
    'continue-and-nudge-both-true-skips',
    {
      events: [],
      respondingParticipantId: P1,
      isContinue: true,
      isNudge: true,
      interval: 5,
      silenceThreshold: 3,
    },
  ],

  // ---- "first" trigger -----------------------------------------------------
  [
    'first-empty-history',
    {
      events: [],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 5,
      silenceThreshold: 3,
    },
  ],
  [
    'first-character-has-spoken-but-never-whispered-bootstrap',
    {
      events: [myTurn(P1), theirTurn(P2), myTurn(P1), theirTurn(P2)],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 2,
      silenceThreshold: 10,
    },
  ],
  [
    'first-other-participant-whispered-not-mine',
    {
      events: [coreWhisperMsg(P2), myTurn(P1)],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 5,
      silenceThreshold: 5,
    },
  ],

  // ---- "periodic" trigger ---------------------------------------------------
  [
    'periodic-exact-interval-reached',
    {
      events: [
        coreWhisperMsg(P1),
        myTurn(P1),
        theirTurn(P2),
        myTurn(P1),
        theirTurn(P2),
        myTurn(P1),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 3,
      silenceThreshold: 100,
    },
  ],
  [
    'periodic-one-below-interval-no-fire',
    {
      events: [coreWhisperMsg(P1), myTurn(P1), theirTurn(P2), myTurn(P1)],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 3,
      silenceThreshold: 100,
    },
  ],
  [
    'periodic-interval-one-fires-on-next-turn',
    {
      events: [coreWhisperMsg(P1), myTurn(P1)],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 1,
      silenceThreshold: 100,
    },
  ],
  [
    'periodic-only-counts-turns-after-whisper-not-before',
    {
      events: [myTurn(P1), myTurn(P1), coreWhisperMsg(P1), myTurn(P1)],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 2,
      silenceThreshold: 100,
    },
  ],
  [
    'periodic-resets-after-second-whisper',
    {
      events: [
        coreWhisperMsg(P1),
        myTurn(P1),
        myTurn(P1),
        coreWhisperMsg(P1), // second whisper resets the counter
        myTurn(P1),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 2,
      silenceThreshold: 100,
    },
  ],

  // ---- "silence" trigger -----------------------------------------------------
  [
    'silence-exact-threshold-reached',
    {
      events: [
        coreWhisperMsg(P1),
        myTurn(P1),
        theirTurn(P2),
        theirTurn(P3),
        userTurn(),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 3,
    },
  ],
  [
    'silence-one-below-threshold-no-fire',
    {
      events: [coreWhisperMsg(P1), myTurn(P1), theirTurn(P2), theirTurn(P3)],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 3,
    },
  ],
  [
    'silence-run-resets-on-my-turn',
    {
      events: [
        coreWhisperMsg(P1),
        theirTurn(P2),
        theirTurn(P3),
        myTurn(P1), // resets silenceRun to 0
        theirTurn(P2),
        theirTurn(P3),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 3,
    },
  ],
  [
    'silence-nonmessage-events-do-not-break-or-extend-run',
    {
      events: [
        coreWhisperMsg(P1),
        theirTurn(P2),
        contextSummary(),
        systemEvent(),
        theirTurn(P3),
        theirTurn(P2),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 3,
    },
  ],

  // ---- visibility exclusions (isVisibleConversationalTurn) -------------------
  [
    'staff-whisper-excluded-from-silence-run',
    {
      events: [
        coreWhisperMsg(P1),
        staffWhisper('host'),
        staffWhisper('concierge'),
        staffWhisper('prospero'),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 3,
    },
  ],
  [
    'silent-message-excluded-from-silence-run',
    {
      events: [coreWhisperMsg(P1), silentMsg(P2), silentMsg(P3), silentMsg(P2)],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 3,
    },
  ],
  [
    'empty-trimmed-content-excluded',
    {
      events: [coreWhisperMsg(P1), emptyContentMsg(P2), emptyContentMsg(P3), emptyContentMsg(P2)],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 3,
    },
  ],
  [
    'private-whisper-not-targeting-me-excluded',
    {
      events: [
        coreWhisperMsg(P1),
        privateWhisper([P2, P3]),
        privateWhisper([P2, P3]),
        privateWhisper([P2, P3]),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 3,
    },
  ],
  [
    'private-whisper-targeting-me-counted-as-not-mine',
    {
      events: [
        coreWhisperMsg(P1),
        privateWhisper([P1, P3], 'aside to you'),
        privateWhisper([P1, P3], 'another aside'),
        privateWhisper([P1, P3], 'third aside'),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 3,
    },
  ],
  [
    'my-own-private-whisper-still-counts-as-mine',
    {
      // A message authored BY me (role ASSISTANT + participantId=me) with
      // targetParticipantIds including me is visible and "mine" -> resets
      // silenceRun and counts toward periodic.
      events: [
        coreWhisperMsg(P1),
        theirTurn(P2),
        theirTurn(P3),
        msg({ role: 'ASSISTANT', participantId: P1, targetParticipantIds: [P1, P2], content: 'whisper from me to us' }),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 2,
    },
  ],
  [
    'empty-target-array-treated-as-public',
    {
      events: [
        coreWhisperMsg(P1),
        privateWhisper([], 'targets empty array = public'),
        privateWhisper([], 'still public'),
        privateWhisper([], 'still public again'),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 3,
    },
  ],
  [
    'mixed-visible-and-excluded-events',
    {
      events: [
        coreWhisperMsg(P1),
        staffWhisper('lantern'),
        silentMsg(P2),
        theirTurn(P2),
        emptyContentMsg(P3),
        theirTurn(P3),
        privateWhisper([P2, P3]),
        theirTurn(P2),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 3,
    },
  ],

  // ---- context-transition trigger --------------------------------------------
  [
    'context-transition-fires-after-whisper',
    {
      events: [
        coreWhisperMsg(P1),
        myTurn(P1),
        librarianFold('rolling-summary'),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 100,
      fireOnContextTransition: true,
    },
  ],
  [
    'context-transition-per-character-summary-kind',
    {
      events: [
        coreWhisperMsg(P1),
        myTurn(P1),
        librarianFold('per-character-summary'),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 100,
      fireOnContextTransition: true,
    },
  ],
  [
    'context-transition-before-last-whisper-does-not-fire',
    {
      events: [
        librarianFold('rolling-summary'),
        coreWhisperMsg(P1), // whisper AFTER the fold -> transition is stale
        myTurn(P1),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 100,
      fireOnContextTransition: true,
    },
  ],
  [
    'context-transition-disabled-by-flag',
    {
      events: [
        coreWhisperMsg(P1),
        myTurn(P1),
        librarianFold('rolling-summary'),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 100,
      fireOnContextTransition: false,
    },
  ],
  [
    'context-transition-loose-fold-substring-no-longer-matches',
    {
      // "folder-created-by-character" contains "fold" as a substring but its
      // kebab segments are ['folder','created','by','character'] — none is
      // 'summary' or 'rolling', so the tightened check correctly skips it.
      events: [
        coreWhisperMsg(P1),
        myTurn(P1),
        librarianFold('folder-created-by-character'),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 100,
      fireOnContextTransition: true,
    },
  ],
  [
    'context-transition-underscore-segment-rolling',
    {
      events: [
        coreWhisperMsg(P1),
        myTurn(P1),
        librarianFold('rolling_fold_event'),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 100,
      fireOnContextTransition: true,
    },
  ],
  [
    'context-transition-not-librarian-sender-ignored',
    {
      events: [
        coreWhisperMsg(P1),
        myTurn(P1),
        msg({ role: 'ASSISTANT', systemSender: 'host', systemKind: 'rolling-summary', content: 'not actually librarian' }),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 100,
      fireOnContextTransition: true,
    },
  ],
  [
    'context-transition-per-character-summary-not-targeting-me-excluded',
    {
      events: [
        coreWhisperMsg(P1),
        myTurn(P1),
        librarianFold('per-character-summary', [P2]),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 100,
      fireOnContextTransition: true,
    },
  ],
  [
    'context-transition-per-character-summary-targeting-me-included',
    {
      events: [
        coreWhisperMsg(P1),
        myTurn(P1),
        librarianFold('per-character-summary', [P1, P2]),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 100,
      fireOnContextTransition: true,
    },
  ],
  [
    'context-transition-picks-latest-not-first',
    {
      events: [
        coreWhisperMsg(P1),
        librarianFold('rolling-summary'),
        myTurn(P1),
        librarianFold('rolling-summary'), // latest one, still after whisper
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 100,
      fireOnContextTransition: true,
    },
  ],
  [
    'context-transition-default-flag-true-when-omitted',
    {
      events: [
        coreWhisperMsg(P1),
        myTurn(P1),
        librarianFold('rolling-summary'),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 100,
      // fireOnContextTransition omitted -> defaults to true in v4.
    },
  ],

  // ---- trigger precedence (periodic/silence checked before context-transition) --
  [
    'periodic-takes-precedence-over-context-transition',
    {
      events: [
        coreWhisperMsg(P1),
        myTurn(P1),
        myTurn(P1),
        librarianFold('rolling-summary'),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 2,
      silenceThreshold: 100,
      fireOnContextTransition: true,
    },
  ],
  [
    'silence-takes-precedence-over-context-transition',
    {
      events: [
        coreWhisperMsg(P1),
        theirTurn(P2),
        theirTurn(P3),
        librarianFold('rolling-summary'),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 100,
      silenceThreshold: 2,
      fireOnContextTransition: true,
    },
  ],

  // ---- no trigger fires -------------------------------------------------------
  [
    'no-trigger-fires-quiet-period',
    {
      events: [coreWhisperMsg(P1), myTurn(P1), theirTurn(P2)],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 5,
      silenceThreshold: 5,
      fireOnContextTransition: true,
    },
  ],
  [
    'no-trigger-fires-with-nonmessage-events-only',
    {
      events: [coreWhisperMsg(P1), contextSummary(), systemEvent()],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 5,
      silenceThreshold: 5,
      fireOnContextTransition: true,
    },
  ],

  // ---- multi-participant realistic scene --------------------------------------
  [
    'realistic-three-participant-scene',
    {
      events: [
        userTurn('Let us begin.'),
        myTurn(P1, 'I greet everyone.'),
        theirTurn(P2, 'Nice to meet you.'),
        coreWhisperMsg(P1),
        myTurn(P1, 'A thoughtful reply.'),
        theirTurn(P3, 'An aside from the third.'),
        userTurn('Continuing on.'),
        theirTurn(P2, 'Another remark.'),
      ],
      respondingParticipantId: P1,
      isContinue: false,
      isNudge: false,
      interval: 3,
      silenceThreshold: 3,
      fireOnContextTransition: true,
    },
  ],
];

for (const [id, opts] of cases) {
  const out = shouldFireCoreWhisper(opts as ShouldFireCoreWhisperOptions);
  rows.push({
    id,
    options: {
      events: opts.events as unknown as Array<Record<string, unknown>>,
      respondingParticipantId: opts.respondingParticipantId,
      isContinue: opts.isContinue,
      isNudge: opts.isNudge,
      interval: opts.interval,
      silenceThreshold: opts.silenceThreshold,
      fireOnContextTransition: opts.fireOnContextTransition,
    },
    out,
  });
}

for (const r of rows) process.stdout.write(JSON.stringify(r) + '\n');

File: crates/quilltap-core/src/core_whisper.rs
//! Aurora Core whisper — cadence trigger.
//!
//! Port of v4's lib/chat/context/core-whisper-trigger.ts. Pure function. Given
//! a chat's event history and the participant about to take the next turn,
//! decide whether to offer this character their own `Core/` packet before
//! that turn fires.
//!
//! The three triggers (composed by OR; only one fire per turn):
//!
//! - **first** — the character has never been offered a Core whisper in this
//!   chat. Covers two cases: the character has not yet spoken (the literal
//!   first-turn case) AND chats that predate the feature (the bootstrap
//!   case, where the character has spoken before but never been offered a
//!   Core packet).
//! - **periodic** — `interval` of their own visible turns have passed since
//!   the last Core whisper for them.
//! - **silence** — the last `silenceThreshold` visible conversational turns
//!   immediately before this one were authored by *someone else* (user or
//!   another participant). Long stretches without your voice are exactly
//!   where convergence pressure builds; the packet says "you're still
//!   here — what do you think?"
//!
//! Visible conversational turn (`isVisibleConversationalTurn`) excludes:
//!
//! - Non-message events (`context-summary`, `system`).
//! - Staff whispers — any message with a `systemSender` set.
//! - `isSilentMessage === true`.
//! - Empty / tool-call-only assistant turns (trimmed content empty).
//! - Private whispers targeted away from this responding character
//!   (`targetParticipantIds` set, but doesn't include them) — those aren't
//!   part of this character's room cadence.
//!
//! Continue / nudge turns skip the whisper entirely — a continuation is not a
//! new response.
//!
//! Computed on demand from the message history (no persistent counter).

/// The subset of a `ChatEvent` this module reads. `event_type` mirrors v4's
/// discriminated `type` field (`"message"` / `"context-summary"` / `"system"`);
/// only `"message"` events carry the rest of the fields.
#[derive(Debug, Clone, PartialEq, Default)]
pub struct WhisperEvent {
    pub event_type: String,
    pub role: Option<String>,
    pub participant_id: Option<String>,
    pub content: Option<String>,
    pub system_sender: Option<String>,
    pub system_kind: Option<String>,
    pub is_silent_message: Option<bool>,
    pub target_participant_ids: Option<Vec<String>>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CoreWhisperReason {
    First,
    Periodic,
    Silence,
    ContextTransition,
}

impl CoreWhisperReason {
    pub fn as_str(&self) -> &'static str {
        match self {
            CoreWhisperReason::First => "first",
            CoreWhisperReason::Periodic => "periodic",
            CoreWhisperReason::Silence => "silence",
            CoreWhisperReason::ContextTransition => "context-transition",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ShouldFireCoreWhisperResult {
    pub fire: bool,
    pub reason: Option<CoreWhisperReason>,
}

pub struct ShouldFireCoreWhisperOptions<'a> {
    pub events: &'a [WhisperEvent],
    pub responding_participant_id: &'a str,
    pub is_continue: bool,
    pub is_nudge: bool,
    pub interval: i64,
    pub silence_threshold: i64,
    /// When true, the first visible turn after a Librarian rolling-summary
    /// fold counts as a `context-transition` fire.
    pub fire_on_context_transition: bool,
}

/// A "visible conversational turn" for the purposes of this responding
/// character's cadence.
fn is_visible_conversational_turn(m: &WhisperEvent, responding_participant_id: &str) -> bool {
    if m.system_sender.is_some() {
        return false;
    }
    if m.is_silent_message == Some(true) {
        return false;
    }
    match &m.content {
        Some(content) if !crate::jsstr::js_trim(content).is_empty() => {}
        _ => return false,
    }
    if let Some(targets) = &m.target_participant_ids {
        if !targets.is_empty() && !targets.iter().any(|t| t == responding_participant_id) {
            return false;
        }
    }
    true
}

/// Conservative v1 definition of a major context transition: the most recent
/// Librarian rolling-summary fold whisper. Returns its index, or `None`.
///
/// The Librarian's rolling-summary `systemKind` values use kebab-case slugs
/// containing the literal words `summary` or `rolling` (e.g. `rolling-summary`,
/// `per-character-summary`). The earlier substring check on `fold` was too
/// loose — it false-matched announcements like `folder-created-by-character`,
/// which represent a character filing a folder, not a memory fold. The
/// tightened check uses kebab-segment membership so `summary` / `rolling`
/// match cleanly without catching unrelated kinds.
fn find_latest_context_transition_index(
    events: &[WhisperEvent],
    responding_participant_id: &str,
) -> Option<usize> {
    for i in (0..events.len()).rev() {
        let m = &events[i];
        if m.event_type != "message" {
            continue;
        }
        if m.system_sender.as_deref() != Some("librarian") {
            continue;
        }
        let Some(system_kind) = &m.system_kind else {
            continue;
        };
        let lowered = system_kind.to_lowercase();
        let segments: Vec<&str> = lowered.split(['-', '_']).collect();
        if !segments.contains(&"summary") && !segments.contains(&"rolling") {
            continue;
        }
        // A per-character Librarian summary that isn't addressed to this
        // character shouldn't count as a transition for them.
        if let Some(targets) = &m.target_participant_ids {
            if !targets.is_empty() && !targets.iter().any(|t| t == responding_participant_id) {
                continue;
            }
        }
        return Some(i);
    }
    None
}

pub fn should_fire_core_whisper(
    options: ShouldFireCoreWhisperOptions<'_>,
) -> ShouldFireCoreWhisperResult {
    let ShouldFireCoreWhisperOptions {
        events,
        responding_participant_id,
        is_continue,
        is_nudge,
        interval,
        silence_threshold,
        fire_on_context_transition,
    } = options;

    if is_continue || is_nudge {
        return ShouldFireCoreWhisperResult {
            fire: false,
            reason: None,
        };
    }

    let mut last_core_whisper_idx: Option<usize> = None;
    let mut my_self_turn_count: i64 = 0;
    let mut my_self_turns_after_whisper: i64 = 0;
    let mut silence_run: i64 = 0;

    for (i, m) in events.iter().enumerate() {
        if m.event_type != "message" {
            // System/context-summary events neither extend nor break the
            // silence run.
            continue;
        }

        if m.system_sender.as_deref() == Some("aurora")
            && m.system_kind.as_deref() == Some("core-whisper")
            && m.target_participant_ids
                .as_ref()
                .map(|t| t.iter().any(|p| p == responding_participant_id))
                .unwrap_or(false)
        {
            last_core_whisper_idx = Some(i);
            my_self_turns_after_whisper = 0;
            // A Core whisper isn't a turn by anyone; don't touch silence_run.
            continue;
        }

        let visible = is_visible_conversational_turn(m, responding_participant_id);
        if !visible {
            continue;
        }

        let is_mine = m.role.as_deref() == Some("ASSISTANT")
            && m.participant_id.as_deref() == Some(responding_participant_id);

        if is_mine {
            my_self_turn_count += 1;
            if let Some(last_idx) = last_core_whisper_idx {
                if i > last_idx {
                    my_self_turns_after_whisper += 1;
                }
            }
            silence_run = 0;
        } else {
            silence_run += 1;
        }
    }
    let _ = my_self_turn_count; // mirrors v4's unused-after-loop local

    // "First" covers both the literal first-turn case AND the bootstrap case
    // for chats that predate the feature — a character who has spoken many
    // times but never been offered a Core packet gets one on their next turn.
    let Some(last_core_whisper_idx) = last_core_whisper_idx else {
        return ShouldFireCoreWhisperResult {
            fire: true,
            reason: Some(CoreWhisperReason::First),
        };
    };

    if my_self_turns_after_whisper >= interval {
        return ShouldFireCoreWhisperResult {
            fire: true,
            reason: Some(CoreWhisperReason::Periodic),
        };
    }

    if silence_run >= silence_threshold {
        return ShouldFireCoreWhisperResult {
            fire: true,
            reason: Some(CoreWhisperReason::Silence),
        };
    }

    if fire_on_context_transition {
        if let Some(transition_idx) =
            find_latest_context_transition_index(events, responding_participant_id)
        {
            if transition_idx > last_core_whisper_idx {
                return ShouldFireCoreWhisperResult {
                    fire: true,
                    reason: Some(CoreWhisperReason::ContextTransition),
                };
            }
        }
    }

    ShouldFireCoreWhisperResult {
        fire: false,
        reason: None,
    }
}
